README
An implementation of the Nelder-Mead simplex method.

Version 1.0

Copyright (c) 1997-present <Michael F. Hutt>

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

Reference
D. J. Wilde and C. S. Beightler, Foundations of Optimization. Englewood Cliffs, N.J., Prentice-Hall, 1967. 

The Nelder-Mead Simplex Method is direct search algorithm that's
useful for non-linear optimization problems. I was researching
optimization of antenna arrarys at one point and implemented several
versions of the Nelder-Mead algorithm.

The testfun example evaluates the Rosenbrock function.

The minimum was found at
1.000000e+00
1.000000e+00
123 Function Evaluations
61 Iterations through program

The maxpower program is an example of using Nelder-Mead for a one
dimensional optimization problem. The problem is to find the value of
a load resistor that will allow maximum power to be delivered to the
load. For this example I used a simple voltage divider circuit
consisting of a single 9V source with two resistors in series (1000
ohm and 470 ohm) with the output taken across the 470 ohm
resistor. From the Thevenin equivalent circuit we know that the
optimium load resistor value for max power transfer should be 319.73
ohms. The source code for this problem can be found here. The program
produces the following output starting with an initial guess of RL=100
ohms:

RL:320.000000 P:0.006474
29 Function Evaluations
13 Iterations through program


ChangeLog.md
### 2016-05-19 v1.1
	* Added example with starting point of {0.0, 0.0}
	* Renamed test example to rosenbrock

### 2016-05-18 v1.0
	* Re-evaluate all vertices after an unsuccessful contraction.
	* Initial release to github, v1.0

### 2011-03-01
	* Added constraints.

### 2007-07-23
	* Fixed memory leak.

### 1999-01-06
	* Modified to conform to the algorithm presented in Margaret H. Wright's paper on Direct Search Methods.

### 1997-11-03
	* Initial build.



testfun.c
/* 
 * Program: testfun.c
 * Author: Michael F. Hutt
 * http://www.mikehutt.com
 * Mar. 1, 2011
 * gcc -Wall -o testfun testfun.c nmsimplex.c -DMACOSX
 * test function for nmsimplex
 *
 */
 
#include "nmsimplex.h"
#include <stdio.h>
#include <math.h>
 
double rosen(double x[])
{
  return (100*((x[1]-x[0]*x[0])*(x[1]-x[0]*x[0]))+(1.0-x[0])*(1.0-x[0]));
}

double round2(double num, int precision)
{
  double rnum;
  int tnum;

  rnum = num*pow(10,precision);
  tnum = (int)(rnum < 0 ? rnum-0.5 : rnum + 0.5);
  rnum = tnum/pow(10,precision);

  return rnum;
}

void my_constraints(double x[], int n)
{
  // round to 2 decimal places
  int i;

  for (i=0; i<n; i++) {
    x[i] = round2(x[i],2);
  }
}
 
int main()
{
  double start[] = {-1.2,1.0};
  double start_null[2];
  double min;
  int i;

  min=simplex(rosen,start,2,1.0e-4,1,my_constraints);
  for (i=0;i<2;i++) {
    printf("%f:%f\n",start[i],min);
  }
  
  min=simplex(rosen,start_null,2,1.0e-4,1,my_constraints);

  for (i=0;i<2;i++) {
    printf("%f:%f\n",start_null[i],min);
  }
  return 0;
}


nmsimplex.c
/* 
 * Program: nmsimplex.c
 * Author : Michael F. Hutt
 * http://www.mikehutt.com
 * Nov. 3, 1997
 *
 * An implementation of the Nelder-Mead simplex method.
 *
 * Copyright (c) 1997 Michael F. Hutt
 * Released under the MIT License
 *
 */

#include "nmsimplex.h"

/* create the initial simplex */
void initialize_simplex(double **v, double start[], double scale, int n)
{
  double pn,qn;   /* values used to create initial simplex */
  int i,j;
	
  pn = scale*(sqrt(n+1)-1+n)/(n*sqrt(2));
  qn = scale*(sqrt(n+1)-1)/(n*sqrt(2));

  for (i=0;i<n;i++) {
    v[0][i] = start[i];
  }
	
  for (i=1;i<=n;i++) {
    for (j=0;j<n;j++) {
      if (i-1 == j) {
	v[i][j] = pn + start[j];
      }
      else {
	v[i][j] = qn + start[j];
      }
    }
  }
}


/* print out the initial values */
void print_initial_simplex(double **v, double *f, int n)
{
  int i,j;
  
  printf("Initial Values\n");
  for (j=0;j<=n;j++) {
    for (i=0;i<n;i++) {
      printf("%f, ",v[j][i]);
    }
    printf("value %f\n",f[j]);
  }
}


/* print out the value at each iteration */
void print_iteration(double **v, double *f, int n, int itr)
{
  int i,j;
  
  printf("Iteration %d\n",itr);
  for (j=0;j<=n;j++) {
    for (i=0;i<n;i++) {
      printf("%f %f\n",v[j][i],f[j]);
    }
  }
}


/* find the index of the largest value */
int vg_index(double *f, int vg, int n)
{
  int j;
  
  for (j=0;j<=n;j++) {
    if (f[j] > f[vg]) {
      vg = j;
    }
  }
  return vg;
}


/* find the index of the smallest value */
int vs_index(double *f, int vs, int n)
{
  int j;
    
  for (j=0;j<=n;j++) {
    if (f[j] < f[vs]) {
      vs = j;
    }
  }
  return vs;
}


/* find the index of the second largest value */
int vh_index(double *f, int vh, int vg, int n)
{
  int j;
  
  for (j=0;j<=n;j++) {
    if (f[j] > f[vh] && f[j] < f[vg]) {
      vh = j;
    }
  }
  return vh;
}


/* calculate the centroid */
void centroid(double *vm, double **v, int n, int vg)
{
  int j,m;
  double cent;
  
  for (j=0;j<=n-1;j++) {
    cent=0.0;
    for (m=0;m<=n;m++) {
      if (m!=vg) {
	cent += v[m][j];
      }
    }
    vm[j] = cent/n;
  }
}


double simplex(double (*objfunc)(double[]), double start[],int n, double EPSILON, double scale, void (*constrain)(double[],int n))
{
	
  int vs;         /* vertex with smallest value */
  int vh;         /* vertex with next smallest value */
  int vg;         /* vertex with largest value */
	
  int i,j,row;
  int k;   	  /* track the number of function evaluations */
  int itr;	  /* track the number of iterations */
	
  double **v;     /* holds vertices of simplex */
  double *f;      /* value of function at each vertex */
  double fr;      /* value of function at reflection point */
  double fe;      /* value of function at expansion point */
  double fc;      /* value of function at contraction point */
  double *vr;     /* reflection - coordinates */
  double *ve;     /* expansion - coordinates */
  double *vc;     /* contraction - coordinates */
  double *vm;     /* centroid - coordinates */
  double min;
	
  double fsum,favg,s;
	
  /* dynamically allocate arrays */
	
  /* allocate the rows of the arrays */
  v =  (double **) malloc ((n+1) * sizeof(double *));
  f =  (double *) malloc ((n+1) * sizeof(double));
  vr = (double *) malloc (n * sizeof(double));
  ve = (double *) malloc (n * sizeof(double));  
  vc = (double *) malloc (n * sizeof(double));  
  vm = (double *) malloc (n * sizeof(double));  
	
  /* allocate the columns of the arrays */
  for (i=0;i<=n;i++) {
    v[i] = (double *) malloc (n * sizeof(double));
  }
	
  /* create the initial simplex */
  initialize_simplex(v,start,scale,n);

  /* impose constraints */
  if (constrain != NULL) {
    for (j=0;j<=n;j++) {
      constrain(v[j],n);
    }
  }
  
  /* find the initial function values */
  for (j=0;j<=n;j++) {
    f[j] = objfunc(v[j]);
  }
  k = n+1;
	
  /* print out the initial values */
  print_initial_simplex(v,f,n);

  /* begin the main loop of the minimization */
  for (itr=1;itr<=MAX_IT;itr++) {     
    /* find the index of the largest value */
    vg = vg_index(f,0,n);

    /* find the index of the smallest value */
    vs = vs_index(f,0,n);
		
    /* find the index of the second largest value */
    vh = vh_index(f,vs,vg,n);
		
    /* calculate the centroid */
    centroid(vm,v,n,vg);

    /* reflect vg to new vertex vr */
    for (j=0;j<=n-1;j++) {
      /*vr[j] = (1+ALPHA)*vm[j] - ALPHA*v[vg][j];*/
      vr[j] = vm[j]+ALPHA*(vm[j]-v[vg][j]);
    }
    if (constrain != NULL) {
      constrain(vr,n);
    }
    fr = objfunc(vr);
    k++;
		
    if (fr < f[vh] && fr >= f[vs]) {
      for (j=0;j<=n-1;j++) {
	v[vg][j] = vr[j];
      }
      f[vg] = fr;
    }
		
    /* investigate a step further in this direction */
    if ( fr <  f[vs]) {
      for (j=0;j<=n-1;j++) {
	/*ve[j] = GAMMA*vr[j] + (1-GAMMA)*vm[j];*/
	ve[j] = vm[j]+GAMMA*(vr[j]-vm[j]);
      }
      if (constrain != NULL) {
        constrain(ve,n);
      }
      fe = objfunc(ve);
      k++;
			
      /* 
	 by making fe < fr as opposed to fe < f[vs], 			   
	 Rosenbrocks function takes 63 iterations as opposed 
	 to 64 when using double variables. 
      */
			
      if (fe < fr) {
	for (j=0;j<=n-1;j++) {
	  v[vg][j] = ve[j];
	}
	f[vg] = fe;
      }
      else {
	for (j=0;j<=n-1;j++) {
	  v[vg][j] = vr[j];
	}
	f[vg] = fr;
      }
    }
		
    /* check to see if a contraction is necessary */
    if (fr >= f[vh]) {
      if (fr < f[vg] && fr >= f[vh]) {
	/* perform outside contraction */
	for (j=0;j<=n-1;j++) {
	  /*vc[j] = BETA*v[vg][j] + (1-BETA)*vm[j];*/
	  vc[j] = vm[j]+BETA*(vr[j]-vm[j]);
	}
	if (constrain != NULL) {
          constrain(vc,n);
        }
	fc = objfunc(vc);
	k++;
      }
      else {
	/* perform inside contraction */
	for (j=0;j<=n-1;j++) {
	  /*vc[j] = BETA*v[vg][j] + (1-BETA)*vm[j];*/
	  vc[j] = vm[j]-BETA*(vm[j]-v[vg][j]);
	}
	if (constrain != NULL) {
          constrain(vc,n);
        }
	fc = objfunc(vc);
	k++;
      }
			
			
      if (fc < f[vg]) {
	for (j=0;j<=n-1;j++) {
	  v[vg][j] = vc[j];
	}
	f[vg] = fc;
      }

      else {
	/* 
	   at this point the contraction is not successful,
	   we must halve the distance from vs to all the 
	   vertices of the simplex and then continue.
	   1997-10-31 - modified to account for ALL vertices. 
	*/
	
	for (row=0;row<=n;row++) {
	  if (row != vs) {
	    for (j=0;j<=n-1;j++) {
	      v[row][j] = v[vs][j]+(v[row][j]-v[vs][j])/2.0;
	    }
	  }
	}

	/* re-evaluate all the vertices */
	for (j=0;j<=n;j++) {
	  f[j] = objfunc(v[j]);
	}
	
	/* find the index of the largest value */
	vg = vg_index(f,0,n);
	
	/* find the index of the smallest value */
	vs = vs_index(f,0,n);
	
	/* find the index of the second largest value */
	vh = vh_index(f,vs,vg,n);

	if (constrain != NULL) {
          constrain(v[vg],n);
        }
	f[vg] = objfunc(v[vg]);
	k++;
	if (constrain != NULL) {
          constrain(v[vh],n);
        }
	f[vh] = objfunc(v[vh]);
	k++;
				
				
      }
    }
		
    /* print out the value at each iteration */
    /*print_iteration(v,f,n,itr);*/
		
    /* test for convergence */
    fsum = 0.0;
    for (j=0;j<=n;j++) {
      fsum += f[j];
    }
    favg = fsum/(n+1);
    s = 0.0;
    for (j=0;j<=n;j++) {
      s += pow((f[j]-favg),2.0)/(n);
    }
    s = sqrt(s);
    if (s < EPSILON) break;
  }
  /* end main loop of the minimization */
	
  /* find the index of the smallest value */
  vs = vs_index(f,0,n);
	
  /*printf("The minimum was found at\n"); */
  for (j=0;j<n;j++) {
    /*printf("%e\n",v[vs][j]);*/
    start[j] = v[vs][j];
  }
  min=objfunc(v[vs]);
  k++;
  printf("%d Function Evaluations\n",k);
  printf("%d Iterations through program\n",itr);
	
  free(f);
  free(vr);
  free(ve);
  free(vc);
  free(vm);
  for (i=0;i<=n;i++) {
    free (v[i]);
  }
  free(v);
  return min;
}







nmsimplex.h
/*
 * header for nmsimplex.c
 * Author : Michael F. Hutt
 * http://www.mikehutt.com
 * 
 * An implementation of the Nelder-Mead simplex method.
 *
 * Copyright (c) 1997 Michael F. Hutt
 * Released under the MIT License
 *
 */

#ifndef NM_SIMPLEX_H
#define NM_SIMPLEX_H

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#ifdef MACOSX
#include <malloc/malloc.h>
#else
#include <malloc.h>
#endif

#define MAX_IT      1000      /* maximum number of iterations */
#define ALPHA       1.0       /* reflection coefficient */
#define BETA        0.5       /* contraction coefficient */
#define GAMMA       2.0       /* expansion coefficient */

void my_constraints(double x[], int n);
void initialize_simplex(double **v, double start[], double scale, int n);
void print_initial_simplex(double **v, double *f, int n);
void print_iteration(double **v, double *f, int n, int itr);
int vg_index(double *f, int vg, int n);
int vs_index(double *f, int vs, int n);
int vh_index(double *f, int vh, int vg, int n);
void centroid(double *vm, double **v, int n, int vg);
double simplex(double (*objfunc)(double[]), double start[],int n, double EPSILON, double scale, void (*constrain)(double[],int n));

#endif


rosenbrock.c
/* 
 * Program: rosenbrock.c
 * Author: Michael F. Hutt
 * http://www.mikehutt.com
 * May 19, 2016
 * gcc -Wall -o rosenbrock rosenbrock.c nmsimplex.c -DMACOSX
 * test function for nmsimplex
 *
 */
 
#include "nmsimplex.h"
#include <stdio.h>
#include <math.h>
 
double rosen(double x[])
{
  return (100*((x[1]-x[0]*x[0])*(x[1]-x[0]*x[0]))+(1.0-x[0])*(1.0-x[0]));
}

double round2(double num, int precision)
{
  double rnum = 0.0;
  int tnum;

  if (num == 0.0)
    return num;

  rnum = num*pow(10,precision);
  tnum = (int)(rnum < 0 ? rnum-0.5 : rnum + 0.5);
  rnum = tnum/pow(10,precision);

  return rnum;
}

void my_constraints(double x[], int n)
{
  // round to 2 decimal places
  int i;

  for (i=0; i<n; i++) {
    x[i] = round2(x[i],2);
  }
}
 
int main()
{
  double start[] = {-1.2,1.0};
  double start_null[] = {0.0,0.0};
  double min;
  int i;

  /* run optimization with a starting point */
  min=simplex(rosen,start,2,1.0e-4,1,my_constraints);
  printf("The minimum was found at\n"); 
  for (i=0;i<2;i++) {
    printf("%0.2f, ",start[i]);
  }
  printf("value %0.2f\n\n",min);

  /* 
     use default starting point of:
     0,0
     0.965,0.259
     0.259,0.965
  */

  min=simplex(rosen,start_null,2,1.0e-4,1,my_constraints);
  printf("The minimum was found at\n"); 
  for (i=0;i<2;i++) {
    printf("%0.2f, ",start_null[i]);
  }
  printf("value %0.2f\n",min);
  
  return 0;
}


maxpower.c
/* 
 * Program: maxpower.c
 * Author: Michael F. Hutt
 * http://www.mikehutt.com
 * Mar. 2, 2011
 * gcc -Wall -o maxpower maxpower.c nmsimplex.c -DMACOSX
 * 
 * Example to test out nelder-mead on a one dimensional optimization problem
 *
 * The circuit used for this example is a voltage divider consisting of a 1k ohm
 * and 470 ohm resistor with a 9V source. The output is take across the 470 ohm resistor.
 *
 * spice file circuit description
 * Vs 1 0 9
 * R1 1 2 1000
 * R2 2 0 470
 * RL 2 0 320
 *
 * The Thevenin equivalent resistance is 319.73 ohms, which is the value we should find
 * from the optimization.
 *
 *
 * RL:320.000000 P:0.006474
 * 29 Function Evaluations
 * 13 Iterations through program
 * 
 */
 
#include "nmsimplex.h"
#include <stdio.h>
#include <math.h>
 
 
double objfun(double x[])
{
  /* find the power for this value of RL */
  double Req;
  double VL;
  double PL;
  double Vs = 9.0;
  double RL = x[0];
   
  Req = 470*RL/(470+RL);
  VL = Vs*Req/(1000+Req);
  PL = VL*VL/RL;
  /*printf("RL:%f P:%f\n",RL,PL);*/
  return -PL; /* nm finds the minimum, so to find the max we return the negative value */
}
 
void my_constraints(double x[], int n)
{
  // resistance must be positive and not equal to zero
  int i;

  for (i=0; i<n; i++) {
    /*if (x[i] < 1.0) {
      x[i] = 1 + i*50;
      }*/
    if (x[i] < 0) {
      x[i] = fabs(x[i]);
    }
  }
}

int main()
{
  double start[] = {100};
  double min;
  int i;
  int dim = 1;
  double eps = 1.0e-8;
  double scale = 1.0;

  min=simplex(objfun,start,dim,eps,scale,my_constraints);

  for (i=0; i<dim; i++) {
    printf("%0.2f ohms, %f watts\n",start[i],fabs(min));
  }
  return 0;
}
